metric_learn/base_metric.py

```diff
@@ -1,12 +1,43 @@
-from numpy.linalg import inv, cholesky
-from sklearn.base import BaseEstimator, TransformerMixin
+from numpy.linalg import cholesky
+from sklearn.base import BaseEstimator
 from sklearn.utils.validation import check_array
 
+class BaseMetricLearner(BaseEstimator):
 
-class BaseMetricLearner(BaseEstimator, TransformerMixin):
   def __init__(self):
     raise NotImplementedError('BaseMetricLearner should not be instantiated')
 
+
+  def fit_transform(self, X, y=None, **fit_params):
+    """Fit to data, then transform it.
+
+    Fits transformer to X and y with optional parameters fit_params
+    and returns a transformed version of X.
+
+    Parameters
+    ----------
+    X : array-like of shape [n_samples, n_features], or ConstrainedDataset
+        Training set.
+
+    y : numpy array of shape [n_samples] or 4-tuple of arrays
```

Comment:
with the new API `y` should only be constraint labels or `None`, shouldn't it?

Reply:
Thanks, indeed I forgot to update the 4-tuple part 
Originally I wanted to allow ``fit_transform`` to be used in a classical way (so with labels as y) but also in the weakly supervised way (with constraints labels) (and it could be None in both cases). So something like     
```python
y : None or numpy array of shape [n_samples] or [n_constraints]
        Target values, or constraints labels
```
However now I am not sure it it a good idea: maybe we should make this ``fit_transform`` specific to weakly supervised learning (so with ``y`` being constraints labels or None as you say) and inline it into ``WeaklySupervisedMixin`` rather than ``BaseEstimator``, and make ``SupervisedMixin`` inherit from scikit-learn's ``TransformerMixin`` to have the classical version of ``fit_transform`` for supervised estimators, like this: 

```python
class WeaklySupervisedMixin(object):
     def fit_transform(self, X_constrained, y=None, **fit_params):
    """
[...]
    y : None, or numpy array of shape [n_constraints]
        Constraints labels.
    """
class SupervisedMixin(TransformerMixin):
```